gamemodes/terrortown/gamemode/server/sv_player_ext.lua

```diff
@@ -789,119 +795,261 @@ local function FindCorpse(ply)
--- @param function check an additional checking @{function}
--- @param boolean needcorpse whether the dead @{Player} CORPSE is needed
--- @param boolean force if the @{Player} revive is already forced. Useful if you have multiple reviving equipments
--- @param function onFail this @{function} is called if the revive fails
--- @realm server
-function plymeta:Revive(delay, fn, check, needcorpse, force, onFail)
+-- @param [default=3]number delay The delay of the revive
+-- @param [opt]function OnRevive The @{function} that should be run if the @{Player} revives
+-- @param [opt]function DoCheck An additional checking @{function}
+-- @param [default=false]boolean needsCorpse Whether the dead @{Player} @{CORPSE} is needed
+-- @param [default=false]boolean blockRound Stops the round from ending if this is set to true until the player is alive again
+-- @param [opt]function OnFail This @{function} is called if the revive fails
+-- @param [opt]Vector spawnPos The position where the player should be spawned, accounts for minor obstacles
+-- @param [opt]Angle spawnEyeAngle The eye angles of the revived players
+-- @realm server
+function plymeta:Revive(delay, OnRevive, DoCheck, needsCorpse, blockRound, OnFail, spawnPos, spawnEyeAngle)
 	local ply = self
 	local name = "TTT2RevivePlayer" .. ply:EntIndex()
 
-	if timer.Exists(name) or ply.reviving then return end
+	if ply:IsReviving() then return end
 
-	if force then
-		ply.forceRevive = true
-	end
+	delay = delay or 3
+
+	ply:SetReviving(true)
+	ply:SetBlockingRevival(blockRound)
+	ply:SetRevivalStartTime(CurTime())
+	ply:SetRevivalDuration(delay)
 
-	ply.reviving = true
+	ply.OnReviveFailedCallback = OnFail
 
 	timer.Create(name, delay, 1, function()
 		if not IsValid(ply) then return end
 
-		ply.forceRevive = nil
-		ply.reviving = nil
+		ply:SetReviving(false)
+		ply:SetBlockingRevival(false)
+		ply:SendRevivalReason(nil)
 
-		if not isfunction(check) or check(ply) then
-			local corpse = FindCorpse(ply)
+		ply.OnReviveFailedCallback = nil
 
-			if needcorpse and (not IsValid(corpse) or corpse:IsOnFire()) then
-				ply:ChatPrint("You have not been revived because your body no longer exists.")
+		if not isfunction(DoCheck) or DoCheck(ply) then
+			local corpse = FindCorpse(ply)
 
-				timer.Remove(name) -- TODO needed?
+			if needsCorpse and (not IsValid(corpse) or corpse:IsOnFire()) then
+				OnReviveFailed(ply, OnFail, "message_revival_failed_missing_body")
 
 				return
 			end
 
 			ply:SpawnForRound(true)
 
-			if IsValid(corpse) then
-				local spawnPos = FindCorpsePosition(corpse)
-				if spawnPos then
-					ply:SetPos(spawnPos)
-					ply:SetEyeAngles(Angle(0, corpse:GetAngles().y, 0))
-				end
+			if not spawnPos and IsValid(corpse) then
+				spawnPos = corpse:GetPos()
+				spawnEyeAngle = Angle(0, corpse:GetAngles().y, 0)
 			end
 
-			hook.Call("PlayerLoadout", GAMEMODE, ply)
+			spawnPos = spawnPos or ply:GetDeathPosition()
+			spawnPos = spawn.MakeSpawnPointSafe(self, spawnPos)
```

Comment:
this can be combined with the previous line

Reply:
You are possibly changing the `spawnPos`, but not the `spawnAngle` if given?

Reply:
> this can be combined with the previous line

Hmm, I don't know how to combine it in a nice way. It would be
```lua
spawnPos = spawnPos or ply:GetDeathPosition() or spawn.MakeSpawnPointSafe(self, spawnPos or ply:GetDeathPosition())
```
wouldn't it?

Reply:
> You are possibly changing the `spawnPos`, but not the `spawnAngle` if given?

And yes, the spawn angle doesn't rely matter. It's more or less just a visual thing, there are no unsafe spawn angles.

Reply:
But if you change the position, wouldn't you have to change the angle as well? :o

Reply:
No, it would be `spawnPos = spawn.MakeSpawnPointSafe(self, spawnPos or ply:GetDeathPosition())`

Reply:
> No, it would be
> 
> ```lua
> spawnPos = spawn.MakeSpawnPointSafe(self, spawnPos or ply:GetDeathPosition())
> ```

Ah yes, I'm stupid!

> But if you change the position, wouldn't you have to change the angle as well? :o

I don't calculate the viewangle. In which way should I "correct" it?

Reply:
Yea I don't know, you are right, there is no data to calc the viewangle :D

Reply:
yes. The angle is either given as parameter or based on the corpse position